### PR TITLE
Account for certificate state when determining priority (EXPOSUREAPP-8559)

### PR DESCRIPTION
**Pull request by d4rken**
When finding the highest priority certificate, take the certificate state (valid/expired/invalid) into account.

First group the certificates by state, then process them in batches.

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/core/PersonCertificatesExtensions.kt [hunk 1]
```diff
@@ -184,61 +183,75 @@ private fun Collection<CwaCovidCertificate>.rule9FindOldRaTest(
...
-        return it
+): CwaCovidCertificate = this
+    .also { Timber.v("findHighestPriorityCertificate(nowUtc=%s): %s", nowUtc, this) }
+    .run {
+        val valid = filter {
+            it.getState() is CwaCovidCertificate.State.Valid || it.getState() is CwaCovidCertificate.State.ExpiringSoon
+        }
+        val expired = filter { it.getState() is CwaCovidCertificate.State.Expired }
+        val invalid = filter { it.getState() is CwaCovidCertificate.State.Invalid }
+        if (this.size != (valid.size + expired.size + invalid.size)) {
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/core/PersonCertificatesExtensions.kt:197]: Wouldn't this be something for a unit test?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/core/PersonCertificatesExtensions.kt [hunk 2]
```diff
@@ -184,61 +183,75 @@ private fun Collection<CwaCovidCertificate>.rule9FindOldRaTest(
...
+            throw IllegalStateException("State grouping failure, certificate count does not match.")
+        }
+        listOf(valid, expired, invalid)
     }
+    .map { certsForState ->
+        certsForState.rule1FindRecentPcrCertificate(nowUtc)?.let {
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/core/PersonCertificatesExtensions.kt:203]: Is this relying on valid being the first in the list?

Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/core/PersonCertificatesExtensions.kt:203]: Maybe add a comment...

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/core/PersonCertificatesExtensions.kt (hunk 1 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/core/PersonCertificatesExtensions.kt:197]: I'm not sure how I could test this.
The idea is to have something that causes a crash if someone adds new states in the future.
I will check again if I can make it an exhaustive `when`.

**Review by chiljamgossow — APPROVED** [review]
🚀 🏄‍♂️

**Review by mtwalli — APPROVED** [review]

**Pull request merged**